src/v/datalake/catalog_schema_manager.cc

```diff
@@ -187,19 +190,75 @@ catalog_schema_manager::ensure_table_schema(
 
     iceberg::transaction txn(std::move(load_res.value()));
 
-    // Check schema compatibility
-    auto type_copy = writer_struct_type.copy();
-    auto get_res = apply_evolution_rules(table_id, txn.table(), type_copy);
-    if (get_res.has_error()) {
-        co_return get_res.error();
+    std::optional<iceberg::struct_type> new_schema;
+    if (use_schema_merging) {
+        const auto* schema = txn.table().get_schema(
+          txn.table().current_schema_id);
+
+        auto writer_struct_type_with_ids = writer_struct_type.copy();
+        auto fill_res = iceberg::try_fill_field_ids(
+          schema->schema_struct, writer_struct_type_with_ids);
+
+        if (fill_res == iceberg::ids_filled::no) {
+            auto merged_schema_struct_type = schema->schema_struct.copy();
+            auto merge_res = iceberg::merge_struct_types(
+              writer_struct_type, merged_schema_struct_type);
+            if (merge_res.has_error()) {
+                vlog(
+                  datalake_log.warn,
+                  "Failed to merge record schema with table schema {}: {}",
+                  table_id,
+                  merge_res.error());
+                co_return errc::not_supported;
+            }
+
+            // Make a copy so that we still have the original struct in case we
+            // need
+            // to log it on the error path.
+            auto merged_schema_with_evo = merged_schema_struct_type.copy();
+            auto get_res = apply_evolution_rules(
+              table_id, txn.table(), merged_schema_with_evo, false);
+            if (get_res.has_error()) {
+                co_return get_res.error();
+            } else if (get_res.value()) {
+                vlog(
+                  datalake_log.error,
+                  "Applying evolution rules on merged schema resulted in a "
+                  "no-op. "
+                  "This means there is a bug in one of: filling field ids, "
+                  "merging, or applying evolution rules. desired_type={}, "
+                  "merged_schema={}, merged_schema_with_evo={}",
+                  table_id,
+                  writer_struct_type,
+                  merged_schema_struct_type,
+                  merged_schema_with_evo);
+                co_return errc::failed;
+            } else {
+                new_schema.emplace(std::move(merged_schema_with_evo));
+            }
+        }
+    } else {
+        // Check schema compatibility
+        auto type_copy = writer_struct_type.copy();
+        auto get_res = apply_evolution_rules(
+          table_id,
```

Comment:
nit: if accept_historical_schema is only true in this one place, consider moving the call to get_equivalent_schema() here?

Reply:
Excellent suggestion. Squashed this into *datalake: use schema merging* commit. Diff https://github.com/redpanda-data/redpanda/compare/9765c8c5a0..d121e805046d57737fedad9f7489143737c07452

 Also added 2 commits to improve readability:
1) https://github.com/redpanda-data/redpanda/pull/27460/commits/dca0fd1fe4cb39b0deec7ec7ccc73a0a5a4f2c5c
2) https://github.com/redpanda-data/redpanda/pull/27460/commits/2aa8de5ba539d799df7d7794f4dad40c1f4201a4

All changes together in force-push https://github.com/redpanda-data/redpanda/compare/9765c8c5a00876914b524dc0681afb5c69aa0926..2aa8de5ba539d799df7d7794f4dad40c1f4201a4